File: memories/2023/solution/scala/day10/src/main/scala/games/FizzBuzz.scala
package games

object FizzBuzz {
  private val MIN = 0
  private val MAX = 100
  private val FIZZ = 3
  private val BUZZ = 5

  def convert(input: Integer): String =
    if (isOutOfRange(input)) throw new OutOfRangeException
    convertSafely(input)

  private def convertSafely(input: Integer): String =
    if (is(FIZZ, input) && is(BUZZ, input)) {
      return "FizzBuzz"
    }
    if (is(FIZZ, input)) {
      return "Fizz"
    }
    if(is(BUZZ, input)) {
      return "Buzz"
    }
    input.toString


  private def is(divisor: Int, input: Int): Boolean = input % divisor == 0

  private def isOutOfRange(input: Int) = input <= MIN || input > MAX
}

File: memories/2023/solution/php/day10/src/FizzBuzz.php
<?php

declare(strict_types = 1);

namespace Games;

class FizzBuzz
{
	private const MIN = 0;

	private const MAX = 100;

	private const FIZZ = 3;

	private const BUZZ = 5;

	private const FIZZBUZZ = 15;

	public function convert(int $input): string | int | OutOfRangeException
	{
		if ($this->isOutOfRange($input)) {
			throw new OutOfRangeException('Input is out of range');
		}

		return $this->convertSafely($input);
	}

	private function convertSafely(int $input): string | int
	{
		return match (true) {
			$this->is(self::FIZZBUZZ, $input) => 'FizzBuzz',
			$this->is(self::FIZZ, $input) => 'Fizz',
			$this->is(self::BUZZ, $input) => 'Buzz',
			default => $input,
		};
	}

	private function is(int $divisor, int $input): bool
	{
		return $input % $divisor === 0;
	}

	private function isOutOfRange(int $input): bool
	{
		return self::MIN >= $input || self::MAX < $input;
	}
}
